src/ir/ranges.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2018 Parsa Amini
+// Copyright (c) 2018 Hartmut Kaiser
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <phylanx/config.hpp>
+#include <phylanx/execution_tree/primitives/base_primitive.hpp>
+#include <phylanx/ir/ranges.hpp>
+
+#include <hpx/include/util.hpp>
+#include <hpx/throw_exception.hpp>
+
+#include <cstdint>
+
+namespace phylanx { namespace ir
+{
+    //////////////////////////////////////////////////////////////////////////
+    execution_tree::primitive_argument_type range_iterator::dereference() const
+    {
+        switch (it_.index())
+        {
+        case 0:    // int_range_type
+            return execution_tree::primitive_argument_type(
+                util::get<0>(it_).first);
+        case 1:    // arg_range_type
+            return *(util::get<1>(it_));
+        }
+
+        HPX_THROW_EXCEPTION(hpx::invalid_status,
+            "phylanx::ir::range_iterator::dereference",
+            "range_iterator object holds unsupported data type");
+    }
+
+    bool range_iterator::equal(range_iterator const& other) const
+    {
```

Comment:
Please make sure both iterators hold the same type (index) before comparing the values themselves.

Reply:
Done.